Sources/BSON/BSONObjectID.swift

```diff
@@ -0,0 +1,89 @@
+import Foundation
+import NIO
+import NIOConcurrencyHelpers
+
+/// A struct to represent the BSON ObjectID type.
+public struct BSONObjectID: Equatable, Hashable, CustomStringConvertible {
+    /// This `BSONObjectID`'s data represented as a `String`.
+    public var hex: String { self.oid.reduce("") { $0 + String(format: "%02x", $1) } }
+
+    public var description: String { self.hex }
+
+    private let oid: [UInt8]
+    private static let generator = ObjectIDGenerator()
+
+    /// Initializes a new `BSONObjectID`.
+    public init() {
+        self.oid = Self.generator.generate()
+    }
+
+    /// Initializes a new `BSONObjectID`.
+    internal init(_ bytes: [UInt8]) {
+        self.oid = bytes
+    }
+
+    /// Initializes an `BSONObjectID` from the provided hex `String`.
+    /// - Throws:
+    ///   - `BSONError.InvalidArgumentError` if string passed is not a valid BSONObjectID
+    /// - SeeAlso: https://github.com/mongodb/specifications/blob/master/source/objectid.rst
+    public init(_ hex: String) throws {
+        guard hex.utf8.count == 24 else {
+            throw BSONError.InvalidArgumentError(message: "Cannot create ObjectId from \(hex)")
+        }
+        let len = hex.utf8.count / 2
+        var data = Data(capacity: len)
+        for i in 0..<len {
+            let j = hex.index(hex.startIndex, offsetBy: i * 2)
+            let k = hex.index(j, offsetBy: 2)
+            let bytes = hex[j..<k]
+            if var num = UInt8(bytes, radix: 16) {
+                data.append(&num, count: 1)
+            } else {
+                throw BSONError.InvalidArgumentError(message: "Cannot create UInt8 from \(bytes)")
+            }
+        }
+        self = BSONObjectID([UInt8](data))
+    }
+}
+
+extension BSONObjectID: BSONValue {
+    internal static var bsonType: BSONType { .objectID }
+
+    internal var bson: BSON { .objectID(self) }
+
+    internal static func read(from buffer: inout ByteBuffer) throws -> BSON {
+        guard let bytes = buffer.readBytes(length: 12) else {
+            throw BSONError.InternalError(message: "Cannot read 12 bytes for BSONObjectID")
+        }
+        return .objectID(BSONObjectID(bytes))
+    }
+
+    internal func write(to buffer: inout ByteBuffer) {
+        buffer.writeBytes(self.oid)
+    }
+}
+
+internal class ObjectIDGenerator {
+    private let randomNumber: [UInt8]
+    private var counter: NIOAtomic<UInt32>
+    internal init() {
+        // 5 bytes of a random number
+        self.randomNumber = Array(withUnsafeBytes(
+            of: UInt64.random(in: 0...0xFF_FFFF_FFFF).bigEndian, [UInt8].init
```

Comment:
here and elsewhere, it's probably a bit clearer to just use the `.max` constants (e.g. `UInt64.max` in this case)

Reply:
I think that would reduce the randomness, not sure though. It's supposed to be a 5 byte number and if I go beyond 5 bytes for the upper limit than that upper limit has the chance to occur many times as opposed to just once, does that make sense?

Reply:
Oh I see. Honestly I just saw a bunch of F's and assumed you were typing the literal for the max value, but I see you were only going for 5 bytes instead of the full 8.